src/storage.rs

```diff
@@ -334,6 +334,138 @@ impl<T: Default> UnprotectedStorage<T> for NullStorage<T> {
     unsafe fn remove(&mut self, _: Index) -> T { Default::default() }
 }
 
+/// Wrapper storage that stores modifications to components in a bitset.
+///
+/// **Note: Never use `.iter()` on a mutable component storage that uses this.**
+///
+/// #Example Usage:
+/// 
+/// ```rust,ignore
+///# extern crate specs;
+///# use specs::{Planner, World, RunArg, Component, TrackedStorage, VecStorage, System, Join};
+/// 
+/// pub struct Comp(u32);
+/// impl Component for Comp {
+///     // `TrackedStorage` acts as a wrapper around another storage.
+///     // You can put any store inside of here (e.g. HashMapStorage, VecStorage, etc.)
+///     type Storage = TrackedStorage<Comp, VecStorage<Comp>>;
+/// }
+/// 
+/// pub struct CompSystem;
+/// impl System<()> for CompSystem {
+///     fn run(&mut self, arg: RunArg, _: ()) {
+///         let (entities, mut comps) = arg.fetch(|w| {
+///             (w.entities(), w.write::<Comp>()) 
+///         });
+/// 
+///         // Iterates over all components like normal.
+///         for (entity, comp) in (&entities, &comps).iter() {
+///             // ...
+///         }
+/// 
+///         // **Never do this**
+///         // This will flag all components as modified regardless of whether the inner loop
+///         // did modify their data.
+///         for (entity, comp) in (&entities, &mut comps).iter() {
+///             // ...
+///         }
+/// 
+///         // Instead do something like:
+///         for (entity, comp) in (&entities, &comps.check()).iter() {
+///             if true { // check whether you should modify this component or not.
+///                 let mut comp = comps.get_mut(entity);
```

Comment:
Hmm, we should be able to make this nicer. Asking for `get_mut` on an entity that we are currently iterating on would be slower than just giving the value right away. Perhaps, the `check()` could iterate over some nice wrapper aka `std::hashmap::Entry`, so that actually accessing the data becomes straightforward (in terms of performance)?